ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,332 @@
+    numeric_cols = set(data._get_numeric_data().columns.values)
+    date_cols = set(data.select_dtypes(include=[np.datetime64]).columns)
+    likely_cat = set(data.columns) - numeric_cols
+    # mypy absolutely looses it if likely_cat is overwritten to be a list
+    likely_cat_no_dates = list(likely_cat - date_cols)
+
+    # check proportion of unique values if numerical
+    for var in data._get_numeric_data().columns:
+        likely_flag = 1.0 * data[var].nunique() / data[var].count() < 0.005
+        if likely_flag:
+            likely_cat_no_dates.append(var)
+    return likely_cat_no_dates
+
+
+class CohortTracker:
+    def __init__(self, adata: AnnData | pd.DataFrame, columns: list = None, categorical: list = None, *args: Any):
+        """Track cohort changes over multiple filtering or processing steps.
+
+        This class offers functionality to track and plot cohort changes over multiple filtering or processing steps,
+        enabling the user to monitor the impact of each step on the cohort.
+
+        Tightly interacting with the `tableone` package [1].
+        Args:
+            adata: :class:`~anndata.AnnData` or :class:`~pandas.DataFrame` object to track.
+            columns: List of columns to track. If `None`, all columns will be tracked.
+            categorical: List of columns that contain categorical variables, if not given will be inferred from the data.
+
+        References
+        ----------
+        [1] Tom Pollard, Alistair E.W. Johnson, Jesse D. Raffa, Roger G. Mark; tableone: An open source Python package for producing summary statistics for research papers, Journal of the American Medical Informatics Association, Volume 24, Issue 2, 1 March 2017, Pages 267–271, https://doi.org/10.1093/jamia/ocw117
+
+        """
+        if isinstance(adata, AnnData):
+            df = adata.obs
+        elif isinstance(adata, pd.DataFrame):
+            df = adata
+        else:
+            raise ValueError("adata must be an AnnData or a DataFrame.")
+
+        self.columns = columns if columns is not None else list(df.columns)
+
+        if columns is not None:
+            _check_columns_exist(df, columns)
+        if categorical is not None:
+            _check_columns_exist(df, categorical)
+            if set(categorical).difference(set(self.columns)):
+                raise ValueError("categorical columns must be in the (selected) columns.")
+
+        self._tracked_steps: int = 0
+        self._tracked_text: list = []
+        self._tracked_operations: list = []
+
+        # if categorical columns specified, use them
+        # else, follow tableone's logic
+        self.categorical = categorical if categorical is not None else _detect_categorical_columns(df[self.columns])
+        self.track = self._get_column_structure(df)
+
+        self._track_backup = copy.deepcopy(self.track)
+
+    def __call__(
+        self, adata: AnnData, label: str = None, operations_done: str = None, *args: Any, **tableone_kwargs: Any
```

Comment:
This is a question that I don't know the answer to but what's the modern/right way of dealing with kwargs?

I've always just provided a `**kwargs` parameter and then added in the docstring "passed to tableone" because I've seen this pattern many times. But it's stupid because it's less explicit than what you're doing now where you clearly name the kwargs parameter such that it's clear where they're passed to.

Reply:
Happy to hear that you dont know the go-to answer to that either haha

never really understood how collecting anything the function doesnt know (including misspellt arguments) and throw it at downstream function is a great idea, but can settle if its the standard

Will check for 3rd advice on how to make this in a modern way

Reply:
Following the recommendation from another channel, a `tableone_kwargs` dict seems to be a reasonable way to go then?

Reply:
Yes!